File: 2020/Day04/Haskell/PassportProcessing-Part1.hs
{-# LANGUAGE OverloadedStrings #-}

import System.IO
import Control.Monad
import qualified Data.Text as Text

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

groupPass:: [[Char]] -> [[Char]]
groupPass remainder
    | null remainder = []
    | otherwise = unwords pass : groupPass (drop (length pass + 1) remainder)
        where pass = takeWhile(not . null) remainder 
    
needed:: [Text.Text]
needed = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let rows = lines contents
    print (length (filter (\p -> all (\n -> Text.isInfixOf n (Text.pack p)) needed) (groupPass rows)))
    hClose handle

File: 2020/Day04/Python/PassportProcessing-Part1.py
import sys
import collections
import itertools
import numpy as np

passports = []

test = """ecl:gry pid:860033327 eyr:2020 hcl:#fffffd
byr:1937 iyr:2017 cid:147 hgt:183cm

iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884
hcl:#cfa07d byr:1929

hcl:#ae17e1 iyr:2013
eyr:2024
ecl:brn pid:760753108 byr:1931
hgt:179cm

hcl:#cfa07d eyr:2025 pid:166559648
iyr:2011 ecl:brn hgt:59in"""

passp = ""
for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
# for line in test.splitlines():
    if len(line) == 0:
        passports.append(passp)
        passp = ""
    else:
        passp += line.replace("\n"," ")
passports.append(passp)

needed = ["byr", "iyr", "eyr", "hgt","hcl","ecl","pid"]
ans = 0

for p in passports:
    for n in needed:
        if n not in p:
            # print(p, n)
            # input()
            break
    else:
        # print(p)
        ans += 1

print(ans)

